js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,211 @@
+    this.params = params;
+
+    this.location = { page };
+    this.pageSize = pageSize;
+  }
+
+  abstract get type(): string;
+
+  public clear() {
+    this.pages = [];
+    m.redraw();
+  }
+
+  public loadPrev(): Promise<void> {
+    if (this.loadingPrev || this.getLocation().page === 1) return Promise.resolve();
+
+    this.loadingPrev = true;
+
+    const page: number = this.getPrevPageNumber();
+
+    return this.loadPage(page)
+      .then(this.parseResults.bind(this, page))
+      .finally(() => (this.loadingPrev = false));
+  }
+
+  public loadNext(): Promise<void> {
+    if (this.loadingNext) return Promise.resolve();
+
+    this.loadingNext = true;
+
+    const page: number = this.getNextPageNumber();
+
+    return this.loadPage(page)
+      .then(this.parseResults.bind(this, page))
+      .finally(() => (this.loadingNext = false));
+  }
+
+  protected parseResults(pageNum, results: T[]) {
+    const page = {
+      number: pageNum,
+      items: results,
+    };
+
+    if (this.isEmpty() || pageNum > this.getNextPageNumber() - 1) {
+      this.pages.push(page);
+    } else {
+      this.pages.unshift(page);
+    }
+
+    this._hasNext = !!results.payload?.links?.next;
+    this._hasPrev = !!results.payload?.links?.prev;
+
+    m.redraw();
+  }
+
+  /**
+   * Load a new page of results.
+   */
+  protected loadPage(page = 1): Promise<T[]> {
+    const params = this.requestParams();
+    params.page = { offset: this.pageSize * (page - 1) };
+
+    if (Array.isArray(params.include)) {
+      params.include = params.include.join(',');
+    }
+
+    return app.store.find(this.type, params);
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request.
+   * Do not include page offset unless subclass overrides loadPage.
+   *
+   * @abstract
+   * @see loadPage
+   */
+  protected requestParams(): any {
+    return this.params;
+  }
+
+  /**
+   * Update the `this.params` object, calling `refresh` if they have changed.
+   * Use `requestParams` for converting `this.params` into API parameters
+   *
+   * @param newParams
+   * @param page
+   * @see requestParams
+   */
+  public refreshParams(newParams, page: number) {
+    if (this.isEmpty() || this.paramsChanged(newParams)) {
+      this.params = newParams;
+
+      return this.refresh(page);
+    }
+  }
+
+  public refresh(page: number = 1) {
+    this.initialLoading = true;
+    this.loadingPrev = false;
+    this.loadingNext = false;
+
+    m.redraw();
+
+    this.location = { page };
+
+    return this.loadPage()
+      .then((results: T[]) => {
+        this.pages = [];
+        this.parseResults(this.location.page, results);
+      })
+      .finally(() => (this.initialLoading = false));
+  }
+
+  public getPages() {
+    return this.pages;
+  }
+  public getLocation(): PaginationLocation {
+    return this.location;
+  }
+
+  public isLoading(): boolean {
+    return this.initialLoading || this.loadingNext;
```

Comment:
Why not prev?

Reply:
The way the method was being used made it be this.

If the UI shows multiple loading spinners, most likely the implementation will include one for *both* initial & prev, and then another for after (or vice versa). I guess it's confusing what the method does and this logic should probably go in the component itself.

Reply:
I changed it so that `isLoading` checks all 3 load states, and the components themselves use the individual ones depending on how they implement initial loading vs loading prev & next.